fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinLAsscomHelper.java

```diff
@@ -0,0 +1,219 @@
+                    && ExpressionUtils.isIntersecting(
+                    topJoinOnClauseConjunct.collect(SlotReference.class::isInstance), cOutputSlots)
+            ) {
+                return false;
+            }
+        }
+        List<Expression> bottomJoinOnClauseConjuncts = ExpressionUtils.extractConjunction(
+                bottomJoinOnClause);
+
+        List<Expression> allOnCondition = Lists.newArrayList();
+        allOnCondition.addAll(topJoinOnClauseConjuncts);
+        allOnCondition.addAll(bottomJoinOnClauseConjuncts);
+
+        List<SlotReference> newBottomJoinSlots = Lists.newArrayList();
+        newBottomJoinSlots.addAll(aOutputSlots);
+        newBottomJoinSlots.addAll(cOutputSlots);
+
+        for (Expression onCondition : allOnCondition) {
+            List<SlotReference> slots = onCondition.collect(SlotReference.class::isInstance);
+            if (new HashSet<>(newBottomJoinSlots).containsAll(slots)) {
+                newBottomJoinOnCondition.add(onCondition);
+            } else {
+                newTopJoinOnCondition.add(onCondition);
+            }
+        }
+
+        // newBottomJoinOnCondition/newTopJoinOnCondition is empty. They are cross join.
+        // Example:
+        // A: col1, col2. B: col2, col3. C: col3, col4
+        // (A & B on A.col2=B.col2) & C on B.col3=C.col3.
+        // (A & B) & C -> (A & C) & B.
+        // (A & C) will be cross join (newBottomJoinOnCondition is empty)
+        if (newBottomJoinOnCondition.isEmpty() || newTopJoinOnCondition.isEmpty()) {
+            return false;
+        }
+
+        return true;
+    }
+
+
+    /**
+     * Get projectExpr of left and right.
+     * Just for project-inside.
+     */
+    private Pair<List<NamedExpression>, List<NamedExpression>> getProjectExprs() {
+        Preconditions.checkArgument(topJoin.left() instanceof LogicalProject);
+        LogicalProject project = (LogicalProject) topJoin.left();
+
+        List<NamedExpression> projectExprs = project.getProjects();
+        List<NamedExpression> newRightProjectExprs = Lists.newArrayList();
+        List<NamedExpression> newLeftProjectExpr = Lists.newArrayList();
+        for (NamedExpression projectExpr : projectExprs) {
+            List<SlotReference> usedSlotRefs = projectExpr.collect(SlotReference.class::isInstance);
+            if (new HashSet<>(bOutputSlots).containsAll(usedSlotRefs)) {
+                newRightProjectExprs.add(projectExpr);
+            } else {
+                newLeftProjectExpr.add(projectExpr);
+            }
+        }
+
+        return new Pair<>(newLeftProjectExpr, newRightProjectExprs);
+    }
+
+
+    private LogicalJoin<GroupPlan, GroupPlan> newBottomJoin() {
+        return new LogicalJoin(
+                bottomJoin.getJoinType(),
+                Optional.of(ExpressionUtils.and(newBottomJoinOnCondition)),
+                a, c);
+    }
+
+    /**
+     * Create topJoin for project-inside.
+     */
+    public LogicalJoin newProjectTopJoin() {
```

Comment:
i think the better code organization is
according whether right project is empty to new a top right child
and then
according whether left project is empty to new top left child
at last
new top node